### ISSUE TITLE
[BE] Get single data by id

### ISSUE THREAD

**Issue by ryuina** [YoungLeeHan/YoungleehanKorean#57]
https://github.com/YoungLeeHan/YoungleehanKorean/blob/7e913c9bf9db725dcd63306a6c780208b57963ea/server/controllers/product.js#L71C1-L81C3

## Get single data by id
Id is the one which is unique and it exists for identification
Please find data by id on single get. 
If "slug"s are unique, then it will be okay. But generally, ID which is automatically generated generally is always indexed in any database.

### Comments
- gpt says
```
Q: find by id vs find by other unique property which one is faster?
A: The speed of finding a record by ID versus finding it by other unique properties depends on several factors, including the database system, the indexing strategy, and the size of the dataset. Generally, finding a record by ID tends to be faster than finding it by other unique properties due to the following reasons:

1. Primary key indexing: In most database systems, the primary key is indexed, which means that the database maintains a data structure that allows for rapid access to records by their IDs. This indexing mechanism is optimized for fast lookups and can significantly speed up retrieval.

2. Direct access: When querying by ID, the database can directly access the record using the indexed primary key. This is a very efficient operation, as it eliminates the need for scanning or searching through the entire dataset.

3. Smaller data size: The primary key is typically a simple, numeric, or sequentially generated value, and therefore it occupies less space in the database compared to other unique properties, which might be longer strings or complex data types. Smaller data size means faster lookup times.

4. Indexing overhead: When searching by other unique properties, the database may need to utilize non-clustered indexes or perform a full table scan, which can introduce additional overhead and slow down the retrieval process.

However, it's essential to consider the overall database design and query patterns. If certain unique properties are queried frequently and have their indexes, the performance difference may be less pronounced. Additionally, database optimization techniques like caching, query tuning, and indexing strategies can also impact the speed of queries.

In conclusion, finding a record by ID is generally faster due to the presence of primary key indexing and the smaller data size. However, the performance difference may vary depending on the specific use case and database setup. It's always a good practice to profile and benchmark your queries in the context of your application to understand the actual performance implications.
```
- another stackoverflow says
  - https://stackoverflow.com/questions/59325634/is-there-a-difference-between-finding-by-primary-key-vs-finding-by-unique-column

**Comment by sweetrain05** [comment]
Ina, thanks for the great suggestion. 
While I was going through the code below, 
I came to wonder how to fetch the single product information in the case of SingleProductView.js.

On this page, we are using params.slug from useParams() to use in findOne().
If we want to use 'id' instead of 'slug', could you guide us how to implement this? 
(How can we pass on the specific product id from the previous page?)

Server-side code:
```
export const read = async (req, res) => {
    try {
        const product = await Product.findOne({ slug: req.params.slug })
            .select("-images")
            .populate("category");

        res.json(product);
    } catch (err) {
        console.log(err);
    }
};

```

Client-side code:
```
const getProduct = (id) => axios.get(`/product/${id}`);

const DescriptionTabLabel = "Description";
const ReviewsTabLabel = "Reviews";
const Tabs = [DescriptionTabLabel, ReviewsTabLabel]

export default function SingleProductView() {
    useScrollToTop();

    // state
    const [product, setProduct] = useState();
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);
    const [currentTab, setCurrentTab] = useState(DescriptionTabLabel);

    // hooks
    const params = useParams();

    // Load product information from DB
    useEffect(() => {
        if (!params.slug || params.slug === "") { 
            return;
        }
        setIsLoading(true);
        getProduct(params.slug)
          .then(({data}) => setProduct(data))
          .catch((err) => {
            setIsError(true);
            console.log(err);
          })
        .finally(() => {
          setIsLoading(false)
          setIsError(false)
        });

    }, [params.slug]);
```

**Comment by ryuina** [comment]
@sweetrain05 
In this case, we can get into SingleProductView with `product._id` as page params
and use page param "id" to call single product

```
// ProductCard.js
...
<Link to={`product/${product?._id}`}>
...
</Link>
```

And then, now `params.slug` is product._id at SingleProductView.js

At back-end, we need to change data quering logic from "find one that has a slug is SLUG" to "find one with id ID"
And also, I recommend you to change router too to describe the params meaning. 
```
// routes/product.js
...
router.get("/product/:id", read);
```

```
// product.js
export const read = async (req, res) => {
    try {
        const product = await Product.findById(req.params.id)
            .select("-images")
            .populate("category");

        res.json(product);
    } catch (err) {
        console.log(err);
    }
};
```

ref: https://masteringjs.io/tutorials/mongoose/find-by-id

**Comment by sweetrain05** [comment]
Thanks for your comment! 
I agree with you personally that using a prime key(_id) is better for finding data and link consistency. 

However, while we were discussing this matter on the meeting, 
@HyangYoung said that we could achieve better SEO if we use slug instead of id in the URL 
and  @younghyunlee22  also wanted to show the product name in the URL. 

In this case, would it be a better choice for us to leave it as :slug or should we change to :_id?
Will there be a critical issue if we choose to use slug on our URL?
